Test that multi_tensor optimizer state buffers match with single_tensor state buffers

mikaylagawarecki opened the pull request:
Add testing for state of multitensor optimizers suggested in #78807 (previously only the equality of model parameters after a few optimizer steps was being tested)

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #81894

mikaylagawarecki committed:
[WIP] Make sure multi_tensor optimizer buffer updates are inplace

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Make sure multi_tensor optimizer buffer updates are inplace"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Test that multi_tensor state buffers match with single_tensor state buffers"




[ghstack-poisoned]

albanD reviewed (commented):
How were the latter param values the same if the state was not?

albanD reviewed (commented):
Also this doesn't fail for Adam? :o

mikaylagawarecki commented on test/test_optim.py:
```diff
@@ -495,10 +495,23 @@ def test_multi_tensor_optimizers(self):
+                mt_p_state = mt_state[mt_p]
+
+                for k in st_p_state:
+                    # FIXME: remove this when https://github.com/pytorch/pytorch/issues/78807 is resolved
```
@albanD this is where it's failing for Adam, but I'm currently skipping it

albanD replied:
Ho read this too quickly.
And why are the params not affected? Is it just that the number of iteration is too small for this to make a visible difference on the params?

mikaylagawarecki replied:
Yes, it seems so, I'll try increasing the num iterations and see whether this fails

mikaylagawarecki replied:
Increasing num_iterations to 300 made some parameter comparisons fail for Adam and AdamW, I am also starting to see errors for NAdam. 

Additionally, comparisons for `exp_avg_sq` and `exp_avg` are starting to fail for these three optimizers. This is odd to me because iiuc, these two buffers are only updated [here](https://github.com/pytorch/pytorch/blob/master/torch/optim/adam.py#L263-L264) using equations like `exp_avg = beta_1 * exp_avg + (1-beta_1) * grad`, so this seems more like some CUDA tolerance errors(?)

I'm currently printing these mismatches in CI to illustrate this and will update this comment with the full list of assertion failures after CI runs

ngimel replied:
Yeah it's probably accuracy errors accumulated over a lot of steps.

mikaylagawarecki commented on test/test_optim.py:
```diff
@@ -495,10 +495,23 @@ def test_multi_tensor_optimizers(self):
+            st_state = state[0]
+            mt_state = state[1]
+            for st_p, mt_p in zip(res[0], res[1]):
+                self.assertEqual(st_p, mt_p, atol=5e-5, rtol=0)
```
@albanD I'm not quite sure why the params are ok, perhaps max_exp_avg_sq's impact on the params is not significant enough to cause differences of more than 5e-5(?)

mikaylagawarecki committed:
Update on "Test that multi_tensor optimizer state buffers match with single_tensor state buffers"

Add testing for state of multitensor optimizers suggested in #78807 (previously only the equality of model parameters after a few optimizer steps was being tested)




[ghstack-poisoned]

crcrpar reviewed (commented):
will we land this after decreasing `kIterations` to avoid CI failures?

mikaylagawarecki commented:
@crcrpar thanks for the ping, will fix this for landing

mikaylagawarecki committed:
Update on "Test that multi_tensor optimizer state buffers match with single_tensor state buffers"

Add testing for state of multitensor optimizers suggested in #78807 (previously only the equality of model parameters after a few optimizer steps was being tested)




[ghstack-poisoned]

albanD reviewed (approved):
Thanks!

mikaylagawarecki commented:
@pytorchbot merge -g
